src/BuiltInTools/dotnet-watch/Internal/FileWatcher.cs

```diff
@@ -132,14 +135,21 @@ void FileChangedCallback(string path, ChangeKind kind)
             return changedFile;
         }
 
-        public static async ValueTask WaitForFileChangeAsync(string path, IReporter reporter, CancellationToken cancellationToken)
+        public static async ValueTask WaitForFileChangeAsync(string filePath, IReporter reporter, CancellationToken cancellationToken)
         {
-            var fileSet = new Dictionary<string, FileItem>() { { path, new FileItem { FilePath = path } } };
+            using var watcher = new FileWatcher(reporter);
 
-            using var watcher = new FileWatcher(fileSet, reporter);
-            await watcher.GetChangedFileAsync(startedWatching: null, cancellationToken);
+            watcher.WatchDirectories([Path.GetDirectoryName(filePath)!]);
 
-            reporter.Output($"File changed: {path}");
+            var fileChange = await watcher.WaitForFileChangeAsync(
+                changeFilter: (path, kind) => path == filePath ? new ChangedFile(new FileItem { FilePath = path }, kind) : null,
+                startedWatching:
+                null, cancellationToken);
```

Comment:
Is it an interesting scenario if the watcher receives an event before the WaitForFileChangeAsync is hooked up?

Reply:
Potentially. I have updated a couple of call sites to pass in `startedWatching` action. This will make tests that are waiting for specific output more robust. 
Other than that, there are time spans during which we ignore/miss file changes. E.g. between the build finishing and when we report that we are waiting for the user to fix a build error. 
That'd be hard to avoid. Technically, change can also occur during the build and we don't know if it was seen by the build and the included in the compilation or not. 
It's more or less best effort.